pkg/controller/cloud/node_controller.go

```diff
@@ -340,23 +341,35 @@ func (cnc *CloudNodeController) initializeNode(ctx context.Context, node *v1.Nod
 				return err
 			}
 		}
+		return nil
+	}); err != nil {
+		utilruntime.HandleError(err)
+		return
+	}
 
-		curNode, err := cnc.kubeClient.CoreV1().Nodes().Get(node.Name, metav1.GetOptions{})
-		if err != nil {
-			return err
-		}
+	curNode, err := cnc.kubeClient.CoreV1().Nodes().Get(node.Name, metav1.GetOptions{})
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("failed to get node %s: %v", node.Name, err))
+		return
+	}
 
-		cloudTaint := getCloudTaint(curNode.Spec.Taints)
-		if cloudTaint == nil {
-			// Node object received from event had the cloud taint but was outdated,
-			// the node has actually already been initialized.
-			return nil
-		}
+	cloudTaint := getCloudTaint(curNode.Spec.Taints)
+	if cloudTaint == nil {
+		// Node object received from event had the cloud taint but was outdated,
+		// the node has actually already been initialized.
+		return
+	}
 
+	type nodeModify func(*v1.Node)
+	var nodeModifyers []nodeModify
+
+	if err := func() error {
 		if curNode.Spec.ProviderID == "" {
 			providerID, err := cloudprovider.GetInstanceProviderID(ctx, cnc.cloud, types.NodeName(curNode.Name))
 			if err == nil {
-				curNode.Spec.ProviderID = providerID
+				nodeModifyers = append(nodeModifyers, func(n *v1.Node) {
```

Comment:
Before this PR, the whole loop would be retried.  That would re-evaluate the conditional above.  I am not SUPER familiar with this, but it seems that it's safe NOT to re-evaluate the conditional, since (if we did lose a race and hit a conflict) this modify function will be idempotent, right?

If so, please comment.  Maybe above, where you declare vars, explain that such modify funcs MUST be safe idempotent in the face of a conflict-and-retry loop

Reply:
In other words, this is tricky enough that I am asking you to make it more obvious by comments, code structure, or both.

Reply:
I did add godocs mentioning why those `modify` funcs exist and what expectations there are. Also, this particular part actually did introduce a change in behavior, by allowing to overwrite an existing providerID, which wasn't possible before. I fixed that by also doing the empty string check in the `modify`.